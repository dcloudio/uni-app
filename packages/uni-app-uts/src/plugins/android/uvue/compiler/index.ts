import { extend } from '@vue/shared'
import { type CompilerError, baseParse } from '@vue/compiler-core'

import { isAppUVueNativeTag } from '@dcloudio/uni-shared'
import {
  createUniVueTransformAssetUrls,
  getBaseNodeTransforms,
  transformTapToClick,
} from '@dcloudio/uni-cli-shared'
import './runtimeHelpers'

import type { CodegenResult, TemplateCompilerOptions } from './options'
import { generate } from './codegen'
import {
  type DirectiveTransform,
  type NodeTransform,
  transform,
} from './transform'
import { transformIf } from './transforms/vIf'
import { transformFor } from './transforms/vFor'
import { transformModel } from './transforms/vModel'
import { transformShow } from './transforms/vShow'
import { transformVText } from './transforms/vText'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'
import { transformOn } from './transforms/vOnWithModifier'
import { transformBind } from './transforms/vBind'
import { transformSlotOutlet } from './transforms/transformSlotOutlet'
import { transformObjectExpression } from './transforms/transformObjectExpression'
import { transformExpression } from './transforms/transformExpression'
import { transformElements } from './transforms/transformElements'
import { transformStyle } from './transforms/transformStyle'
import { transformVHtml } from './transforms/vHtml'
import { transformMemo } from './transforms/vMemo'
import { transformOnce } from './transforms/vOnce'
import {
  type RawSourceMap,
  SourceMapConsumer,
  SourceMapGenerator,
} from 'source-map-js'
import { trackSlotScopes, trackVForSlotScopes } from './transforms/vSlot'
import { transformElement } from './transforms/transformElement'
import {
  createAssetUrlTransformWithOptions,
  normalizeOptions,
} from '../sfc/compiler/template/transformAssetUrl'
import { createSrcsetTransformWithOptions } from '../sfc/compiler/template/transformSrcset'

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean
): TransformPreset {
  return [
    [
      transformOnce,
      transformIf,
      transformMemo,
      transformFor,
      // order is important
      trackVForSlotScopes,
      transformVHtml,
      transformExpression,
      transformSlotOutlet,
      transformElement,
      trackSlotScopes,
      transformText,
      transformTapToClick,
      transformInterpolation,
      transformObjectExpression,
      transformElements,
      transformStyle,
    ] as any,
    {
      on: transformOn,
      bind: transformBind,
      model: transformModel,
      show: transformShow,
      text: transformVText,
    } as any,
  ]
}

export function compile(
  template: string,
  options: TemplateCompilerOptions
): CodegenResult {
  options.rootDir = options.rootDir || ''
  options.targetLanguage = options.targetLanguage || 'kotlin'
  options.prefixIdentifiers =
    'prefixIdentifiers' in options
      ? options.prefixIdentifiers
      : options.mode === 'module'

  wrapOptionsLog(template, options)

  const isNativeTag =
    options?.isNativeTag ||
    function (tag: string) {
      return (
        isAppUVueNativeTag(tag) ||
        !!options.parseUTSComponent?.(tag, options.targetLanguage!)
      )
    }
  const ast = baseParse(template, {
    comments: false,
    isNativeTag,
    onError: options.onError,
  })
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(
    options.prefixIdentifiers
  )

  // 重要不能传入base
  const assetOptions = normalizeOptions(createUniVueTransformAssetUrls(''))

  transform(
    ast,
    extend({}, options, {
      nodeTransforms: [
        ...nodeTransforms,
        createAssetUrlTransformWithOptions(assetOptions),
        createSrcsetTransformWithOptions(assetOptions),
        ...getBaseNodeTransforms('/'),
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      ),
    })
  )

  const result = generate(ast, options)

  // inMap should be the map produced by ./parse.ts which is a simple line-only
  // mapping. If it is present, we need to adjust the final map and errors to
  // reflect the original line numbers.
  if (options.inMap) {
    if (options.sourceMap) {
      result.map = mapLines(options.inMap, result.map!)
    }
    // if (result.errors.length) {
    //   patchErrors(errors, source, inMap)
    // }
  }

  return result
}

function mapLines(oldMap: RawSourceMap, newMap: RawSourceMap): RawSourceMap {
  if (!oldMap) return newMap
  if (!newMap) return oldMap

  const oldMapConsumer = new SourceMapConsumer(oldMap)
  const newMapConsumer = new SourceMapConsumer(newMap)
  const mergedMapGenerator = new SourceMapGenerator()

  newMapConsumer.eachMapping((m) => {
    if (m.originalLine == null) {
      return
    }

    const origPosInOldMap = oldMapConsumer.originalPositionFor({
      line: m.originalLine,
      column: m.originalColumn ?? 0,
    })

    if (origPosInOldMap.source == null) {
      return
    }

    mergedMapGenerator.addMapping({
      generated: {
        line: m.generatedLine,
        column: m.generatedColumn,
      },
      original: {
        line: origPosInOldMap.line, // map line
        // use current column, since the oldMap produced by @vue/compiler-sfc
        // does not
        column: m.originalColumn ?? 0,
      },
      source: origPosInOldMap.source,
      name: origPosInOldMap.name,
    })
  })

  // source-map's type definition is incomplete
  const generator = mergedMapGenerator as any
  ;(oldMapConsumer as any).sources.forEach((sourceFile: string) => {
    generator._sources.add(sourceFile)
    const sourceContent = oldMapConsumer.sourceContentFor(sourceFile)
    if (sourceContent != null) {
      mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
    }
  })

  generator._sourceRoot = oldMap.sourceRoot
  generator._file = oldMap.file
  return generator.toJSON()
}

function wrapOptionsLog(source: string, options: TemplateCompilerOptions) {
  const { onWarn, onError, inMap } = options
  if (inMap && inMap.sourcesContent?.length) {
    if (onWarn || onError) {
      const originalSource = inMap.sourcesContent![0]
      const offset = originalSource.indexOf(source)
      const lineOffset =
        originalSource.slice(0, offset).split(/\r?\n/).length - 1

      if (onWarn) {
        options.onWarn = (err: CompilerError) => {
          patchError(err, lineOffset, offset)
          onWarn(err)
        }
      }
      if (onError) {
        options.onError = (err: CompilerError) => {
          patchError(err, lineOffset, offset)
          onError(err)
        }
      }
    }
  }
}

function patchError(err: CompilerError, lineOffset: number, offset: number) {
  if (err.loc) {
    err.loc.start.line += lineOffset
    err.loc.start.offset += offset
    if (err.loc.end !== err.loc.start) {
      err.loc.end.line += lineOffset
      err.loc.end.offset += offset
    }
  }
}
