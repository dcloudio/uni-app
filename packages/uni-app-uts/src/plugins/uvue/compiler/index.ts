import { extend } from '@vue/shared'
import {
  baseParse,
  trackSlotScopes,
  trackVForSlotScopes,
  transformElement,
  transformExpression,
} from '@vue/compiler-core'

import { isAppUVueNativeTag } from '@dcloudio/uni-shared'
import { transformTapToClick } from '@dcloudio/uni-cli-shared'
import './runtimeHelpers'

import { CodegenResult, CompilerOptions } from './options'
import { generate } from './codegen'
import { DirectiveTransform, NodeTransform, transform } from './transform'
import { transformIf } from './transforms/vIf'
import { transformFor } from './transforms/vFor'
import { transformModel } from './transforms/vModel'
import { transformShow } from './transforms/vShow'
import { transformVText } from './transforms/vText'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'
import { transformOn } from './transforms/vOn'
import { transformBind } from './transforms/vBind'
import { transformSlotOutlet } from './transforms/transformSlotOutlet'

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean
): TransformPreset {
  return [
    [
      transformIf,
      transformFor,
      // order is important
      trackVForSlotScopes,
      transformExpression,
      transformSlotOutlet,
      transformElement,
      trackSlotScopes,
      transformText,
      transformTapToClick,
      transformInterpolation,
    ] as any,
    {
      on: transformOn,
      bind: transformBind,
      model: transformModel,
      show: transformShow,
      text: transformVText,
    } as any,
  ]
}

export function compile(
  template: string,
  options: CompilerOptions
): CodegenResult {
  const ast = baseParse(template, {
    isNativeTag(tag) {
      return isAppUVueNativeTag(tag)
    },
  })
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(
    options.prefixIdentifiers
  )

  transform(
    ast,
    extend({}, options, {
      prefixIdentifiers: options.prefixIdentifiers,
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      ),
    })
  )

  return generate(ast, options)
}
