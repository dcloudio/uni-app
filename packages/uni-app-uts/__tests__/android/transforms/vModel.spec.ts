import { extend } from '@vue/shared'
import {
  BindingTypes,
  type CompilerOptions,
  type ComponentNode,
  type ElementNode,
  ErrorCodes,
  type ForNode,
  NORMALIZE_PROPS,
  NodeTypes,
  type ObjectExpression,
  type PlainElementNode,
  type VNodeCall,
  baseParse as parse,
  trackSlotScopes,
  transform,
  transformElement,
} from '@vue/compiler-core'
import { transformFor } from '../../../src/plugins/android/uvue/compiler/transforms/vFor'
import { transformExpression } from '../../../src/plugins/android/uvue/compiler/transforms/transformExpression'
import { transformModel } from '../../../src/plugins/android/uvue/compiler/transforms/vModel'
import { generate } from '../../../src/plugins/android/uvue/compiler/codegen'
import type { CallExpression } from '@babel/types'
import { assert } from '../testUtils'

function parseWithVModel(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)

  transform(
    ast,
    extend({}, options, {
      prefixIdentifiers: options.prefixIdentifiers,
      nodeTransforms: [
        transformFor,
        transformExpression,
        transformElement,
        trackSlotScopes,
      ],
      directiveTransforms: { model: transformModel },
      expressionPlugins: ['typescript'],
    })
  )

  return ast
}

describe('compiler: transform v-model', () => {
  test('simple expression', () => {
    assert(
      `<input v-model="model" />`,
      `_cE("input", _uM({
  modelValue: _ctx.model,
  onInput: ($event: UniInputEvent) => {(_ctx.model) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])`
    )
  })

  // #2426
  test('simple expression (with multilines)', () => {
    assert(
      `<input v-model="\nmodel.\nfoo\n" />`,
      `_cE("input", _uM({
  modelValue: \n_ctx.model.
foo
,
  onInput: ($event: UniInputEvent) => {(
_ctx.model.
foo
) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])`
    )
  })

  test('compound expression', () => {
    assert(
      `<input v-model="model[index]" />`,
      `_cE(\"input\", _uM({
  modelValue: _ctx.model[_ctx.index],
  onInput: ($event: UniInputEvent) => {(_ctx.model[_ctx.index]) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, [\"modelValue\", \"onInput\"])`
    )
  })

  test('with argument', () => {
    assert(
      `<Foo v-model:title="model" />`,
      `_cV(_component_Foo, _uM({
  title: _ctx.model,
  "onUpdate:title": $event => {(_ctx.model) = $event}
}), null, 8 /* PROPS */, ["title", "onUpdate:title"])`
    )
  })

  test('with dynamic argument', () => {
    assert(
      `<Foo v-model:[value]="model" />`,
      `_cV(_component_Foo, normalizeProps(_uM({
  [_ctx.value]: _ctx.model,
  ["onUpdate:" + _ctx.value]: $event => {(_ctx.model) = $event}
})), null, 16 /* FULL_PROPS */)`
    )
  })
  test('with modifier lazy', () => {
    assert(
      `<input v-model.lazy="model" />`,
      `_cE(\"input\", _uM({
  modelValue: _ctx.model,
  onBlur: ($event: UniInputBlurEvent) => {(_ctx.model) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, [\"modelValue\", \"onBlur\"])`
    )
  })
  test('with modifier number', () => {
    assert(
      `<input v-model.number="model" />`,
      `_cE(\"input\", _uM({
  modelValue: _ctx.model,
  onInput: ($event: UniInputEvent) => {(_ctx.model) = looseToNumber($event.detail.value)}
}), null, 40 /* PROPS, NEED_HYDRATION */, [\"modelValue\", \"onInput\"])`
    )
  })
  test('with modifier trim', () => {
    assert(
      `<input v-model.trim="model" />`,
      `_cE(\"input\", _uM({
  modelValue: _ctx.model,
  onInput: ($event: UniInputEvent) => {(_ctx.model) = $event.detail.value.trim()}
}), null, 40 /* PROPS, NEED_HYDRATION */, [\"modelValue\", \"onInput\"])`
    )
  })
  test('expression width type', () => {
    assert(
      `<Foo v-model="model as string" />`,
      `_cV(_component_Foo, _uM({
  modelValue: _ctx.model,
  "onUpdate:modelValue": ($event: string) => {(_ctx.model) = $event}
}), null, 8 /* PROPS */, [\"modelValue\", \"onUpdate:modelValue\"])`
    )
    assert(
      `<Foo v-model="(model as string)" />`,
      `_cV(_component_Foo, _uM({
  modelValue: _ctx.model,
  "onUpdate:modelValue": ($event: string) => {(_ctx.model) = $event}
}), null, 8 /* PROPS */, [\"modelValue\", \"onUpdate:modelValue\"])`
    )
    assert(
      `<Foo v-model=" (model as string) " />`,
      `_cV(_component_Foo, _uM({
  modelValue: _ctx.model,
  "onUpdate:modelValue": ($event: string) => {(_ctx.model) = $event}
}), null, 8 /* PROPS */, [\"modelValue\", \"onUpdate:modelValue\"])`
    )
  })
  test('complex expressions wrapped in ()', () => {
    assert(
      `<my-input v-model="(obj.str as string)" />`,
      `_cV(_component_my_input, _uM({
  modelValue: _ctx.obj.str,
  \"onUpdate:modelValue\": ($event: string) => {(_ctx.obj.str) = $event}
}), null, 8 /* PROPS */, [\"modelValue\", \"onUpdate:modelValue\"])`
    )
    assert(
      `<my-input v-model="(obj['t'+i] as string)" />`,
      `_cV(_component_my_input, _uM({
  modelValue: _ctx.obj['t'+_ctx.i],
  \"onUpdate:modelValue\": ($event: string) => {(_ctx.obj['t'+_ctx.i]) = $event}
}), null, 8 /* PROPS */, [\"modelValue\", \"onUpdate:modelValue\"])`
    )
  })

  test('simple expression', () => {
    const root = parseWithVModel('<input v-model="model" />')
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties

    expect(props[0]).toMatchObject({
      key: {
        content: 'modelValue',
        isStatic: true,
      },
      value: {
        content: 'model',
        isStatic: false,
      },
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onInput',
        isStatic: true,
      },
      value: {
        children: [
          '($event: UniInputEvent) => {(',
          {
            content: 'model',
            isStatic: false,
          },
          `) = $event.detail.value}`,
        ],
      },
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('simple expression (with prefixIdentifiers)', () => {
    const root = parseWithVModel('<input v-model="model" />', {
      prefixIdentifiers: true,
    })
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties

    expect(props[0]).toMatchObject({
      key: {
        content: 'modelValue',
        isStatic: true,
      },
      value: {
        content: '_ctx.model',
        isStatic: false,
      },
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onInput',
        isStatic: true,
      },
      value: {
        children: [
          '($event: UniInputEvent) => {(',
          {
            content: '_ctx.model',
            isStatic: false,
          },
          `) = $event.detail.value}`,
        ],
      },
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  // #2426
  test('simple expression (with multilines)', () => {
    const root = parseWithVModel('<input v-model="\n model\n.\nfoo \n" />')
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties

    expect(props[0]).toMatchObject({
      key: {
        content: 'modelValue',
        isStatic: true,
      },
      value: {
        content: '\n model\n.\nfoo \n',
        isStatic: false,
      },
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onInput',
        isStatic: true,
      },
      value: {
        children: [
          '($event: UniInputEvent) => {(',
          {
            content: '\n model\n.\nfoo \n',
            isStatic: false,
          },
          `) = $event.detail.value}`,
        ],
      },
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('compound expression', () => {
    const root = parseWithVModel('<input v-model="model[index]" />')
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties

    expect(props[0]).toMatchObject({
      key: {
        content: 'modelValue',
        isStatic: true,
      },
      value: {
        content: 'model[index]',
        isStatic: false,
      },
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onInput',
        isStatic: true,
      },
      value: {
        children: [
          '($event: UniInputEvent) => {(',
          {
            content: 'model[index]',
            isStatic: false,
          },
          `) = $event.detail.value}`,
        ],
      },
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('compound expression (with prefixIdentifiers)', () => {
    const root = parseWithVModel('<input v-model="model[index]" />', {
      prefixIdentifiers: true,
    })
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties

    expect(props[0]).toMatchObject({
      key: {
        content: 'modelValue',
        isStatic: true,
      },
      value: {
        children: [
          {
            content: '_ctx.model',
            isStatic: false,
          },
          '[',
          {
            content: '_ctx.index',
            isStatic: false,
          },
          ']',
        ],
      },
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onInput',
        isStatic: true,
      },
      value: {
        children: [
          '($event: UniInputEvent) => {(',
          {
            children: [
              {
                content: '_ctx.model',
                isStatic: false,
              },
              '[',
              {
                content: '_ctx.index',
                isStatic: false,
              },
              ']',
            ],
          },
          `) = $event.detail.value}`,
        ],
      },
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('with argument', () => {
    const root = parseWithVModel('<input v-model:foo-value="model" />')
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties
    expect(props[0]).toMatchObject({
      key: {
        content: 'foo-value',
        isStatic: true,
      },
      value: {
        content: 'model',
        isStatic: false,
      },
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onUpdate:fooValue',
        isStatic: true,
      },
      value: {
        children: [
          '($event: UniInputEvent) => {(',
          {
            content: 'model',
            isStatic: false,
          },
          `) = $event.detail.value}`,
        ],
      },
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('with dynamic argument', () => {
    const root = parseWithVModel('<Foo v-model:[value]="model" />')
    const node = root.children[0] as ElementNode
    const props = (node.codegenNode as VNodeCall)
      .props as unknown as CallExpression

    expect(props).toMatchObject({
      type: NodeTypes.JS_CALL_EXPRESSION,
      callee: NORMALIZE_PROPS,
      arguments: [
        {
          type: NodeTypes.JS_OBJECT_EXPRESSION,
          properties: [
            {
              key: {
                content: 'value',
                isStatic: false,
              },
              value: {
                content: 'model',
                isStatic: false,
              },
            },
            {
              key: {
                children: [
                  '"onUpdate:" + ',
                  {
                    content: 'value',
                    isStatic: false,
                  },
                ],
              },
              value: {
                children: [
                  '$event => {(',
                  {
                    content: 'model',
                    isStatic: false,
                  },
                  `) = $event}`,
                ],
              },
            },
          ],
        },
      ],
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('with dynamic argument (with prefixIdentifiers)', () => {
    const root = parseWithVModel('<input v-model:[value]="model" />', {
      prefixIdentifiers: true,
    })
    const node = root.children[0] as ElementNode
    const props = (node.codegenNode as VNodeCall)
      .props as unknown as CallExpression

    expect(props).toMatchObject({
      type: NodeTypes.JS_CALL_EXPRESSION,
      callee: NORMALIZE_PROPS,
      arguments: [
        {
          type: NodeTypes.JS_OBJECT_EXPRESSION,
          properties: [
            {
              key: {
                content: '_ctx.value',
                isStatic: false,
              },
              value: {
                content: '_ctx.model',
                isStatic: false,
              },
            },
            {
              key: {
                children: [
                  '"onUpdate:" + ',
                  {
                    content: '_ctx.value',
                    isStatic: false,
                  },
                ],
              },
              value: {
                children: [
                  '($event: UniInputEvent) => {(',
                  {
                    content: '_ctx.model',
                    isStatic: false,
                  },
                  `) = $event.detail.value}`,
                ],
              },
            },
          ],
        },
      ],
    })

    expect(generate(root, {} as any).code).toMatchSnapshot()
  })

  test('should cache update handler w/ cacheHandlers: true', () => {
    const root = parseWithVModel('<input v-model="foo" />', {
      prefixIdentifiers: true,
      cacheHandlers: true,
    })
    expect(root.cached).toBe(1)
    const codegen = (root.children[0] as PlainElementNode)
      .codegenNode as VNodeCall
    // should not list cached prop in dynamicProps
    expect(codegen.dynamicProps).toBe(`["modelValue"]`)
    expect((codegen.props as ObjectExpression).properties[1].value.type).toBe(
      NodeTypes.JS_CACHE_EXPRESSION
    )
  })

  test('should not cache update handler if it refers v-for scope variables', () => {
    const root = parseWithVModel(
      '<input v-for="i in list" v-model="foo[i]" />',
      {
        prefixIdentifiers: true,
        cacheHandlers: true,
      }
    )
    expect(root.cached).toBe(0)
    const codegen = (
      (root.children[0] as ForNode).children[0] as PlainElementNode
    ).codegenNode as VNodeCall
    expect(codegen.dynamicProps).toBe(`["modelValue", "onInput"]`)
    expect(
      (codegen.props as ObjectExpression).properties[1].value.type
    ).not.toBe(NodeTypes.JS_CACHE_EXPRESSION)
  })

  test('should not cache update handler if it inside v-once', () => {
    const root = parseWithVModel(
      '<view v-once><input v-model="foo" /></view>',
      {
        prefixIdentifiers: true,
        cacheHandlers: true,
      }
    )
    expect(root.cached).not.toBe(2)
    expect(root.cached).toBe(1)
  })

  test('should mark update handler dynamic if it refers slot scope variables', () => {
    const root = parseWithVModel(
      '<Comp v-slot="{ foo }"><input v-model="foo.bar"/></Comp>',
      {
        prefixIdentifiers: true,
      }
    )
    const codegen = (
      (root.children[0] as ComponentNode).children[0] as PlainElementNode
    ).codegenNode as VNodeCall
    expect(codegen.dynamicProps).toBe(`["modelValue", "onInput"]`)
  })

  test('should generate modelModifiers for component v-model', () => {
    const root = parseWithVModel('<Comp v-model.trim.bar-baz="foo" />', {
      prefixIdentifiers: true,
    })
    const vnodeCall = (root.children[0] as ComponentNode)
      .codegenNode as VNodeCall
    // props
    expect(vnodeCall.props).toMatchObject({
      properties: [
        { key: { content: `modelValue` } },
        { key: { content: `onUpdate:modelValue` } },
        {
          key: { content: 'modelModifiers' },
          value: {
            content: `{ trim: true, "bar-baz": true }`,
            isStatic: false,
          },
        },
      ],
    })
    // should NOT include modelModifiers in dynamicPropNames because it's never
    // gonna change
    expect(vnodeCall.dynamicProps).toBe(`["modelValue", "onUpdate:modelValue"]`)
  })

  test('should generate modelModifiers for component v-model with arguments', () => {
    const root = parseWithVModel(
      '<Comp v-model:foo.trim="foo" v-model:bar.number="bar" />',
      {
        prefixIdentifiers: true,
      }
    )
    const vnodeCall = (root.children[0] as ComponentNode)
      .codegenNode as VNodeCall
    // props
    expect(vnodeCall.props).toMatchObject({
      properties: [
        { key: { content: `foo` } },
        { key: { content: `onUpdate:foo` } },
        {
          key: { content: 'fooModifiers' },
          value: { content: `{ trim: true }`, isStatic: false },
        },
        { key: { content: `bar` } },
        { key: { content: `onUpdate:bar` } },
        {
          key: { content: 'barModifiers' },
          value: { content: `{ number: true }`, isStatic: false },
        },
      ],
    })
    // should NOT include modelModifiers in dynamicPropNames because it's never
    // gonna change
    expect(vnodeCall.dynamicProps).toBe(
      `["foo", "onUpdate:foo", "bar", "onUpdate:bar"]`
    )
  })

  test('complex expression', () => {
    const root = parseWithVModel(`<Comp v-model="values['t'+i] as string" />`, {
      prefixIdentifiers: true,
      bindingMetadata: {
        values: BindingTypes.SETUP_CONST,
      },
    })
    const vnodeCall = (root.children[0] as ComponentNode)
      .codegenNode as VNodeCall
    // props
    expect(vnodeCall.props).toMatchObject({
      properties: [
        {
          key: { content: `modelValue` },
          value: { children: [{ content: `_ctx.values['t'+_ctx.i]` }] },
        },
        { key: { content: `onUpdate:modelValue` } },
      ],
    })
  })

  describe('errors', () => {
    test('missing expression', () => {
      const onError = jest.fn()
      parseWithVModel('<span v-model />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: ErrorCodes.X_V_MODEL_NO_EXPRESSION,
        })
      )
    })

    test('empty expression', () => {
      const onError = jest.fn()
      parseWithVModel('<span v-model="" />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: ErrorCodes.X_V_MODEL_MALFORMED_EXPRESSION,
        })
      )
    })

    test('mal-formed expression', () => {
      const onError = jest.fn()
      parseWithVModel('<span v-model="a + b" />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: ErrorCodes.X_V_MODEL_MALFORMED_EXPRESSION,
        })
      )
    })

    test('allow unicode', () => {
      const onError = jest.fn()
      parseWithVModel('<span v-model="变.量" />', { onError })

      expect(onError).toHaveBeenCalledTimes(0)
    })

    test('used on scope variable', () => {
      const onError = jest.fn()
      parseWithVModel('<span v-for="i in list" v-model="i" />', {
        onError,
        prefixIdentifiers: true,
      })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: ErrorCodes.X_V_MODEL_ON_SCOPE_VARIABLE,
        })
      )
    })

    test('used on props', () => {
      const onError = jest.fn()
      parseWithVModel('<view v-model="p" />', {
        onError,
        bindingMetadata: {
          p: BindingTypes.PROPS,
        },
      })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: ErrorCodes.X_V_MODEL_ON_PROPS,
        })
      )
    })
  })
})
