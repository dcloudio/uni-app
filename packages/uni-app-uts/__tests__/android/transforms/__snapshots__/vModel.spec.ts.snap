// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: transform v-model compound expression (with prefixIdentifiers) 1`] = `
"createElementVNode("input", utsMapOf({
  modelValue: _ctx.model[_ctx.index],
  onInput: ($event: UniInputEvent) => {(_ctx.model[_ctx.index]) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])"
`;

exports[`compiler: transform v-model compound expression 1`] = `
"createElementVNode("input", utsMapOf({
  modelValue: model[index],
  onInput: ($event: UniInputEvent) => {(model[index]) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])"
`;

exports[`compiler: transform v-model simple expression (with multilines) 1`] = `
"createElementVNode("input", utsMapOf({
  modelValue: 
 model
.
foo 
,
  onInput: ($event: UniInputEvent) => {(
 model
.
foo 
) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])"
`;

exports[`compiler: transform v-model simple expression (with prefixIdentifiers) 1`] = `
"createElementVNode("input", utsMapOf({
  modelValue: _ctx.model,
  onInput: ($event: UniInputEvent) => {(_ctx.model) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])"
`;

exports[`compiler: transform v-model simple expression 1`] = `
"createElementVNode("input", utsMapOf({
  modelValue: model,
  onInput: ($event: UniInputEvent) => {(model) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["modelValue", "onInput"])"
`;

exports[`compiler: transform v-model with argument 1`] = `
"createElementVNode("input", utsMapOf({
  "foo-value": model,
  "onUpdate:fooValue": ($event: UniInputEvent) => {(model) = $event.detail.value}
}), null, 40 /* PROPS, NEED_HYDRATION */, ["foo-value", "onUpdate:fooValue"])"
`;

exports[`compiler: transform v-model with dynamic argument (with prefixIdentifiers) 1`] = `
"createElementVNode("input", normalizeProps(utsMapOf({
  [_ctx.value]: _ctx.model,
  ["onUpdate:" + _ctx.value]: ($event: UniInputEvent) => {(_ctx.model) = $event.detail.value}
})), null, 16 /* FULL_PROPS */)"
`;

exports[`compiler: transform v-model with dynamic argument 1`] = `
"createVNode(_component_Foo, normalizeProps(utsMapOf({
  [value]: model,
  ["onUpdate:" + value]: $event => {(model) = $event}
})), null, 16 /* FULL_PROPS */)"
`;
